iceoryx_posh/include/iceoryx_posh/internal/roudi/service_registry.hpp

```diff
@@ -32,23 +35,44 @@ static const capro::IdString_t Wildcard{"*"};
 class ServiceRegistry
 {
   public:
-    static constexpr uint32_t MAX_INSTANCES_PER_SERVICE = 100u;
-    using InstanceSet_t = cxx::vector<capro::IdString_t, MAX_INSTANCES_PER_SERVICE>;
-    struct instance_t
+    enum class Error
     {
-        InstanceSet_t instanceSet;
+        INVALID_STATE,
+        SERVICE_REGISTRY_FULL,
     };
-    using serviceMap_t = std::map<capro::IdString_t, instance_t>;
+    /// @todo #415 should be connected with iox::MAX_NUMBER_OF_SERVICES
+    static constexpr uint32_t MAX_SERVICE_DESCRIPTIONS = 100U;
+    using ReferenceCounter_t = uint64_t;
+    using ServiceDescriptionVector_t =
+        cxx::vector<cxx::pair<capro::ServiceDescription, ReferenceCounter_t>, MAX_SERVICE_DESCRIPTIONS>;
 
-    void add(const capro::IdString_t& service, const capro::IdString_t& instance);
-    void remove(const capro::IdString_t& service, const capro::IdString_t& instance);
-    void find(InstanceSet_t& instances,
-              const capro::IdString_t& service,
-              const capro::IdString_t& instance = Wildcard) const;
-    const serviceMap_t& getServiceMap() const;
+    /// @brief Adds given service description to registry
+    /// @param[in] serviceDescription, service to be added
+    /// @return ServiceRegistryError, error wrapped in cxx::expected
+    cxx::expected<Error> add(const capro::ServiceDescription& serviceDescription) noexcept;
+
+    /// @brief Removes given service description from registry
+    /// @param[in] serviceDescription, service to be removed
+    /// @return true, if service description was removed, false otherwise
+    bool remove(const capro::ServiceDescription& serviceDescription) noexcept;
+
+    /// @brief Removes given service description from registry
+    /// @param[in] searchResult, reference to the vector which will be filled with the results
+    /// @param[in] service, string or wildcard to search for
+    /// @param[in] service, string or wildcard to search for
+    void find(ServiceDescriptionVector_t& searchResult,
+              const capro::IdString_t& service = Wildcard,
+              const capro::IdString_t& instinstanceance = Wildcard) const noexcept;
```

Comment:
Is after this PR another PR pending? I am asking since we are providing here only service and instance and event is missing.
If the service registry is aware of the full service description we cannot use this feature for the search. This makes the PR for me incomplete - except when there is another PR pending, then forget my gibberish.

Reply:
@MatthiasKillat is right, this API is modeled after ara::com from AUTOSAR. Agree, this can be added later but currently out of scope for #415 see [DDD](https://github.com/eclipse-iceoryx/iceoryx/blob/master/doc/design/service-discovery.md)